lib/Extractor/Solver.cpp

```diff
@@ -50,27 +62,122 @@ class BaseSolver : public Solver {
-      bool IsSat;
-      EC = SMTSolver->isSatisfiable(BuildQuery(BPCs, PCs, Mapping, 0),
-                                    IsSat, 0, 0, Timeout);
-      if (EC)
-        return EC;
-      if (!IsSat) {
-        RHS = I;
-        return EC;
+
+    if (LHS->Width == 1 && InferI1) {
+      std::vector<Inst *>Guesses { IC.getConst(APInt(1, true)),
+                                   IC.getConst(APInt(1, false)) };
+      for (auto G : Guesses) {
+        // TODO: we can trivially synthesize an i1 undef by checking for
+        // validity of both guesses
+        InstMapping Mapping(LHS, G);
+        bool IsSat;
+        EC = SMTSolver->isSatisfiable(BuildQuery(BPCs, PCs, Mapping, 0),
+                                      IsSat, 0, 0, Timeout);
+        if (EC)
+          return EC;
+        if (!IsSat) {
+          RHS = G;
+          return EC;
+        }
       }
     }
+
+    // TODO: constant synthesis goes here
+
+    // TODO drop inputs that are guaranteed to fail the dominance test
+    std::set<Inst *> Inputs;
+    if (InferNop || InferUnary)
+      for (auto Op : LHS->Ops)
+        getInputs(Op, Inputs);
+
+    if (InferNop) {
+      for (auto I : Inputs) {
+        if (I->Width == 1 &&
+            (I->K == Inst::Const || I->K == Inst::UntypedConst))
+          continue;
+        if (LHS->Width != I->Width)
+          continue;
+        InstMapping Mapping(LHS, I);
+        bool IsSat;
+        EC = SMTSolver->isSatisfiable(BuildQuery(BPCs, PCs, Mapping, 0), IsSat,
+                                      0, 0, Timeout);
+        if (EC)
+          return EC;
+        if (!IsSat) {
+          RHS = I;
+          return EC;
+        }
+      }
+    }
+
+    if (InferUnary) {
+      int LHSCost = cost(LHS);
+      for (auto I : Inputs) {
+        if (I->Width == 1 &&
+            (I->K == Inst::Const || I->K == Inst::UntypedConst))
+          continue;
+        std::vector<Inst *> Guesses;
+        if (LHS->Width > I->Width) {
+          Guesses.push_back(IC.getInst(Inst::SExt, LHS->Width, {I}));
+          Guesses.push_back(IC.getInst(Inst::ZExt, LHS->Width, {I}));
+        } else if (LHS->Width < I->Width) {
+          Guesses.push_back(IC.getInst(Inst::Trunc, LHS->Width, {I}));
+        } else {
+          Guesses.push_back(IC.getInst(Inst::Xor, LHS->Width,
+              { IC.getConst(APInt(I->Width, -1)), I }));
+          Guesses.push_back(IC.getInst(Inst::Sub, LHS->Width,
+              { IC.getConst(APInt(I->Width, 0)), I }));
+          if (I->Width == 16 || I->Width == 32 || I->Width == 64)
+            Guesses.push_back(IC.getInst(Inst::BSwap, LHS->Width, {I}));
+          if (I->Width == 8 || I->Width == 16 || I->Width == 32 ||
+              I->Width == 64 || I->Width == 256) {
+            Guesses.push_back(IC.getInst(Inst::CtPop, LHS->Width, {I}));
+            Guesses.push_back(IC.getInst(Inst::Cttz, LHS->Width, {I}));
+            Guesses.push_back(IC.getInst(Inst::Ctlz, LHS->Width, {I}));
```

Comment:
Definitely not, but keep in mind that this code only runs when the user asks for it and also it's intended to be provisional and will get ripped out when we have a real synthesis strategy.